src/MatchesSolver.php

```diff
@@ -46,7 +46,7 @@ protected function result(
         array $matrix
     ) {
         return array_map(function (Match $result) use ($stringA) {
-            $result->value = substr($stringA, $result->index(), $result->length);
+            $result->value = mb_substr($stringA, $result->index(), $result->length);
```

Comment:
`mb_substr()` is lacking the optional fourth argument. The string will be assumed to be encoded using whatever happens to be the return value of `mb_internal_encoding()`. This may or may not be the encoding that the string uses.

We can try to get the correct encoding using `mb_detect_encoding()`. But, hey, that can get tricky since the return value might be `false`. Oh, and `mb_detect_encoding()` can get things wrong if a string has invalidly-encoded characters which is why we need to also set the optional third argument to `true`. Doing so has the (very useful) side benefit of significantly reducing the chance of an invalid encoding attack.

Not entirely sure how you'd work this into these changes, but here's an example of the logic you'd need to use:

```php
function detectMultibyteStringEncoding(string $string): string
{
    $encoding = mb_detect_encoding($string, mb_detect_order(), true);

    // Return the detected encoding or 'utf-8' if not detected. 
    // This is just an example, ideally 'utf-8' would be defined in a well-named constant.
    return is_string($encoding) ? $encoding : 'utf-8';
}

// ...

$result->value = mb_substr($stringA, $result->index(), $result->length, $encoding);
```

The same applies to all calls to `mb_*()` functions that can accept an optional `$encoding` argument, of which there are some more in these set of changes.

Aren't mb strings fun!